### ISSUE TITLE
Bug: [no-unused-vars] false positive when using module augmentation

### ISSUE THREAD

**Issue by jakubmazanec** [typescript-eslint/typescript-eslint#10518]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.4&fileType=.tsx&code=JYWwDg9gTgLgBAbwLACg5xgTzAUzgJQgHcARAQxjIBpUBfOAMyghDgHIABSgOwGdKAxgGsA9FBxkBMALSUARgBscbANypUAExwCFZcXBAQNAVyXsuZPoNHjJM%2BUraJU6YNxg4oDSXgDCEBWMQbgBZHEoAHgAVcko4HAAPD24NXgJiWOo4KIA1MkCcAD5nNHRGYAUPKDyoYEsYAH4ALnYoSwBzZTgAH3ZeHCUpJ162DyTVFzhaOiA&eslintrc=N4KABGBEAOCGBOBnApvSAuKABALgT2mUQGN4BLaHAWiIBsyA7HAejiVUgBpwp4BXWkQxhQECJFwEipCtTqMWDAPZU%2BDPigAmVAG4JEwgNo8xUVPCVpupiMEh6kwyLFq0uUBAHMDmSMobI7pBknsrwyABKRDgAymQARvQM3sI4-MgmYgC%2BmQC6PDk5QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
import {
  type RowData,
} from '@tanstack/react-table';

declare module '@tanstack/react-table' {
  // type parameters are required because `ColumnMeta` must have the same type parameters as in the original module
  interface ColumnMeta<TData extends RowData, TValue> { // ESLint error here
    filterVariant?: 'range' | 'select' | 'text';
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    '@typescript-eslint/no-unused-vars': [
      'error',
      {vars: 'all', args: 'none', ignoreRestSiblings: true},
    ],
  },
};
```

### Expected Result

No error.

### Actual Result

Error. And if I remove the type parameters ESLint is complaining about, TypeScript errors: `All declarations of 'ColumnMeta' must have identical type parameters`.
